engine/src/main/java/org/terasology/engine/physics/bullet/BulletPhysics.java

```diff
@@ -330,133 +548,216 @@ public void update(float delta) {
         } catch (Exception e) {
             logger.error("Error running simulation step.", e);
         }
-        collisions.addAll(getNewCollisionPairs());
-    }
 
-    public btDiscreteDynamicsWorld getDiscreteDynamicsWorld() {
-        return this.discreteDynamicsWorld;
+        PerformanceMonitor.startActivity("Physics Update");
+        updateCollisionPairs();
+        updateRigidBodies(delta);
+        if (networkSystem.getMode().isServer() && time.getGameTimeInMs() - TIME_BETWEEN_NETSYNCS > lastNetsync) {
+            sendSyncMessages();
+            lastNetsync = time.getGameTimeInMs();
+        }
+        PerformanceMonitor.endActivity();
     }
 
-    @Override
-    public boolean removeRigidBody(EntityRef entity) {
-        BulletRigidBody rigidBody = entityRigidBodies.remove(entity);
-        if (rigidBody != null) {
-            removeRigidBody(rigidBody);
-            // wake up this entities neighbors
+    private void updateRigidBodies(float delta) {
+        Matrix4f transform = new Matrix4f();
```

Comment:
We should definitely avoid allocating new instances of things in the `update(delta)` calls whenever possible. Same probably holds for all the vectors allocated here.